plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,217 @@
+#include <caffe/proto/caffe.pb.h>
+#include <dmlc/logging.h>
+#include <dmlc/parameter.h>
+#include <mxnet/operator.h>
+#include <atomic>
+
+#include "../../src/operator/operator_common.h"
+#include "caffe_common.h"
+#include "caffe_stream.h"
+#include "caffe_fieldentry.h"
+#include "caffe_blob.h"
+#include "../../src/io/inst_vector.h"
+#include "../../src/io/iter_prefetcher.h"
+#include "../../src/operator/cast-inl.h"
+
+namespace mxnet {
+namespace io {
+
+struct CaffeDataParam : public dmlc::Parameter<CaffeDataParam> {
+  /*! \brief protobuf text */
+  ::caffe::LayerParameter prototxt;
+  /*! \brief maximum iteration count */
+  int max_iterations;
+  /*! \brief data mode */
+  bool flat;
+
+  DMLC_DECLARE_PARAMETER(CaffeDataParam) {
+    DMLC_DECLARE_FIELD(prototxt).set_default("layer{}")
+      .describe("Caffe's layer parameter");
+    DMLC_DECLARE_FIELD(flat).set_default(false)
+      .describe("Augmentation Param: Whether to flat the data into 1D.");
+    DMLC_DECLARE_FIELD(max_iterations).set_lower_bound(1).set_default(10000)
+      .describe("Maximum number of iterations per epoch.");
+  }
+};
+
+struct CaffePrefetchParam : dmlc::Parameter<CaffePrefetchParam> {
+    /*! \brief data type */
+    int dtype;
+
+    DMLC_DECLARE_PARAMETER(CaffePrefetchParam) {
+      DMLC_DECLARE_FIELD(dtype)
+        .add_enum("float32", mshadow::kFloat32)
+        .add_enum("float64", mshadow::kFloat64)
+        .add_enum("float16", mshadow::kFloat16)
+        .add_enum("uint8", mshadow::kUint8)
+        .add_enum("int32", mshadow::kInt32)
+        .set_default(mshadow::default_type_flag)
+        .describe("Data type.");
+    }
+};
+
+template<typename Dtype>
+class CaffeDataIter : public IIterator<TBlobBatch> {
+ public:
+  CaffeDataIter(void) : loc_(0), batch_size_(0), channels_(0), width_(0), height_(0) {}
+  virtual ~CaffeDataIter(void) {}
+
+  // intialize iterator loads data in
+  virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+    std::map<std::string, std::string> kmap(kwargs.begin(), kwargs.end());
+    param_.InitAllowUnknown(kmap);
+
+    // Caffe seems to understand phase inside an "include {}" block
+    if (!param_.prototxt.has_phase()) {
+      if (param_.prototxt.include().size()) {
+        if (param_.prototxt.include(0).has_phase()) {
+          param_.prototxt.set_phase(param_.prototxt.include(0).phase());
+        }
+      }
+    }
+
+    std::string type = param_.prototxt.type();
+    caffe_data_layer_ = caffe::LayerRegistry<Dtype>::CreateLayer(param_.prototxt);
+    if (caffe_data_layer_) {
+      const size_t top_size = param_.prototxt.top_size();
+      if (top_size > 0) {
+        top_.reserve(top_size);
+        for (size_t x = 0; x < top_size; ++x) {
+          ::caffe::Blob<Dtype> *blob = new ::caffe::Blob<Dtype>();
+          cleanup_blobs_.push_back(std::unique_ptr<::caffe::Blob<Dtype>>(blob));
+          top_.push_back(blob);
+        }
+        caffe_data_layer_->SetUp(bottom_, top_);
```

Comment:
Ok. Note similar Cast parameter takes all.of these. Any idea why?
On Mon, Aug 22, 2016 at 6:19 PM Haoran Wang <notifications@github.com>
wrote:

> In plugin/caffe/caffe_data_iter.cc
> <https://github.com/dmlc/mxnet/pull/3103#discussion_r75789068>:
>
> > +        }
> > +      }
> > +    }
> > +
> > +    std::string type = param_.prototxt.type();
> > +    caffe_data_layer_ = caffe::LayerRegistry<Dtype>::CreateLayer(param_.prototxt);
> > +    if (caffe_data_layer_) {
> > +      const size_t top_size = param_.prototxt.top_size();
> > +      if (top_size > 0) {
> > +        top_.reserve(top_size);
> > +        for (size_t x = 0; x < top_size; ++x) {
> > +          ::caffe::Blob<Dtype> *blob = new ::caffe::Blob<Dtype>();
> > +          cleanup_blobs_.push_back(std::unique_ptr<::caffe::Blob<Dtype>>(blob));
> > +          top_.push_back(blob);
> > +        }
> > +        caffe_data_layer_->SetUp(bottom_, top_);
>
> Yes. I think caffe could handle float and double.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dmlc/mxnet/pull/3103/files/88b4ac0671371b8ec7e1516b64f183251c4a8efa#r75789068>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKts_YZVFZh9VNyO2coTY8qz1ZTxOSw8ks5qikqXgaJpZM4JqHuN>
> .
>